### ISSUE TITLE
Test on "big datasets" and "heavy RF configurations": PLS selection, memory occupancy, running time and global results

### ISSUE THREAD

**Issue by aestoup** [diyabc/abcranger#25]
During the week 24-30 of June I did run two tests (LINUX environment) of the LAST RELEASE of program estimParam on "big pop-genetics datasets" using "heavy RF configurations" to: (i) see PLS selection outputs as we are now using a new "clean" procedure, (ii) see how memory was occupied / managed, (iii) estimate computation times and (iv) see if results were globally sensible when compare with a (much) smaller configuration treated using abcRF in R.

Dataset 1: Harmonia axyridis "S17-moustaches" - NGS:
- 11796 summary statistics
- 13000 particules in the training dataset
- 3000 trees in the forest
- parameter of interest: ra7
- no test dataset required as we will use out-of-bag tricks to do the job.
Conclusions: 
- 720 PLS components selected, which seems quite big too me cf. only 134 PLS axes have importance weights > NOISE...but if the procedure is clean let's keep it (as at the end noise variables are managed by RF) ...but optimally can we use the same clean PLS selection procedure in a more "stringent" way to reduce the number of selected PLS axes? Note: that this large(r) number of selected PLS axes (compared to more basic / less clean  procedure) is also observed on many other datasets.
- Computation duration = about 3 hours on my computer (with 32 cores).
- Memory: about 21 Go at max....which is OK on my computer (cf. the latter has 63 Go of memory) but which might be (too) large on many other computers. So my question is: can we decrease memory occupancy? = for example by running on a smaller number of cores. Is it proportional = if I use 16 instead of 32 cores will the mermory occupancy be decreased by a factor 2? Will the computation time be doubled in return?
- Final results (importance weights of statistics, prior error rate, mean / quantiles) seem OK.

Dataset 2 - Pygmees - microsatellites
- 204 summary statistics
- 100000 particules in the training dataset
- 3000 trees in the forest
- parameter of interest: tccp
- no test dataset required as we will use out-of-bag tricks to do the job.
Conclusions: 
- 51 PLS components selected, which seems quite big too me cf. only 15 PLS axes have importance weights > NOISE...but if the procedure is clean let's keep it (as at the end noise variables are managed by RF) ...but optimally can we use the same clean PLS selection procedure in a more "stringent" way to reduce the number of selected PLS axes? 
- Computation duration = about 30 min on my computer (with 32 cores).
- Memory: about 23 Go stable at max (small initial, pick at 42 GO due to the reading of all reftable.bin wich contain 800000 datasets !!!)....which is OK on my computer (cf. the latter has 63 Go of memory) but which might be (too) large on many other computers. ----> same question as for dataset 1.
- Final results (importance weights of statistics, prior error rate, mean / quantiles) seem OK.

**Comment by fradav** [comment]
The only way to reduce memory consumption for the moment is to reduce the number of threads (-j option). There is currently no workaround about the memory footprint of a big reftable like the Dataset 2. 

I think there is two ways of overcoming big reftable : 

- the "mmap" way, we map the reftable file in memory but we "read" it only  as needed. There is a huge architectural impact to the code in order to make the threads efficiently constructs trees while reading the same chunks of the reftable, so we avoid frequent file-system reads. I don't know if it is even possible
- the "online" way, we dig the traditional breiman's Random Forests for "online" versions, which can train forests while reading reftable sequentially. It isn't our RF anymore there is a lot of different outcomes to be expected (and for example, certain online RF algorithm have a much higher sensibility to noise then breiman's RF)

Another question we may ask to our ABC expert : are ~1M lines long reftable really useful ? Is there a limit over which any additional simulated particles become irrelevant ?

**Comment by fradav** [comment]
Two questions for you 
you said : 
> no test dataset required as we will use out-of-bag tricks to do the job.

How is that possible? Currently the OOB replacement (at least in EstimParam) for test evaluation is still NOT implemented. 

> Final results (importance weights of statistics, prior error rate, mean / quantiles) seem OK

You didn't observe the variance drift wrt abcrf I disccused there #24 ?

**Comment by aestoup** [comment]
Answer to copmment 1: (sorry in French)

J'aurais probablement du ecrire : We will use "in a close futur" ... Je n'ai pas activé dans mes tests la partie calculs sur "tests datatest" because ca prends bc de temps et comme in fine on utilisera OOB ca n'avait a priori pas de sens de voir ce que les calculs sur "tests datatest" donnait en terme de temps de calcul et charge memoire...Je referai des tests " temps de calcul et charge memoire" en incluant des calculs d'erreur/de precision une fois que l'option OOB (a la Louis; voir le ms Chapuis et al bioarXiv en PJ - notamment Supplementary Material S1) sera implémenté (ceci aussi bien pour modelChoice que pour estimParam).

J'imagine que tu as deja discuté de ca avec Jean-Michel et/ou Louis?

Answer to comment 2: (sorry in French!)
Pas vraiment observable sur un seul run test basé sur un dataobs!

Ceci dit j'ai vu tes graphes #24. Je suis daccord  avec ton interpretation = variance differentes pour abcranger et abcrf (c'etait au depart ca que je signalais sur des bases certes bc moins robustes donc c'est bien d'avoir quantifié ca comme ca)

Pour la moyenne ca semble ok.

Quid de la mediane (qui est une estimation ponctuelle que on utilise bc)?

Quid de quantile 5% et quantile 95% (encore des mesures ponctuelles que on utilise bc pour donner des intervales de credibilité)

Dans tous les cas il faut effectivement tenter de comprendre pourquoi tu as ces differences sur la variance.

Malheureusement c'est quelque chose ou je ne peux pas trop aider = voir avec Jean-Michel et/ou Louis peut etre du coté des estimateurs de variance ou autres trucs. Du coup je n'ai pas specialement réagi. Desolé.

PS: il manque peut etre dans tes messages GitUb sur ce sujet qques elements de contexte et d'interpretation = quel jeu de donné utilises tu + combien de test datasets pour faire tes distributions ?

**Comment by aestoup** [comment]
Answer to the question "Another question we may ask to our ABC expert : are ~1M lines long reftable really useful ? Is there a limit over which any additional simulated particles become irrelevant ?"

Indeed 1M is much too big = I do believe that 100000 datasets of a GIVEN scenario in the reftable is by far large enough for parameter estimation under such scenario...but users may be tempted to use a reftable with simulated data from several different scenarios...for instance 10 different scenarios cf. a reftable previously used for model choice. 10 x 100000 = 1M. So in that context if one wants to make inferences from 100000 datasets for parameter x under a single scenario z we will need to read all 1M datasets in the reftable...

**Comment by fradav** [comment]
Indeed, even if the chosen scenario restricts the dataset by an order of magnitude, we load the whole original dataset to filter out the chosen scenario. But I can make an ad hoc reftable reader which only stores particles corresponding to the chosen scenario for exactly this purpose. I propose it for milestone 1.2

**Comment by fradav** [comment]
Pour la question OOB vs test : oui j'ai déjà discuté de ça avec Louis qui est OK, JMM a répondu qu'il y réfléchira (peut-être le relancer?). Ca ne concerne que l'estimation de paramètres, pour le choix de modèle, c'est déjà fait. Le problème étant qu'on ne peut pas utiliser tout l'OOB pour le test en estimparam, car les calculs de poids sont trop coûteux, on garde le paramètre de --test mais au lieu de splitter en train/test, on prend --ntest comme nombre d'OOB sur lesquel on va calculer les prédictions/poids. 
Est-ce que c'est plus clair ?

**Comment by aestoup** [comment]
"I can make an ad hoc reftable reader which only stores particles corresponding to the chosen scenario for exactly this purpose. I propose it for milestone 1.2"
---> great! I like that.

**Comment by aestoup** [comment]
OOB: new stat methodology to compute two types of errors = global prior errors and Local posterior errors

For details see Chapuis et al bioarXiv (send by email to you) - Supplementary Material S1: Overview of the used ABC Random Forest (ABC-RF) methods - especially p 4 and 5

**Comment by fradav** [comment]
I don't understand what has changed in this paper relatively to the reference papers from pubdlo/raynal. All errors global/local are computed with OOB, except for the time being, global error in estim param.

**Comment by fradav** [comment]
Also, an easy workaround for big reftable generated only to have enough particles in the selected scenario:  first generate a reasonable/normal amount of particles for model choice, and _once_ the right scenario is chosen, modify header, remove every scenarii but the this one and finally generate the reftable for estim-param. What do you think ?

PS : variance drift fixed, see #24

**Comment by aestoup** [comment]
This is indeed a possibility and I personally often concretely do that.

...But, according to my experience, some people will not do this additional simulation effort (especially whjen simulation are very costly).

So one can suggest/encourage to do that in "the notice", but we have to implement a memory non-costly reftable reading process that just load the particules of the scenario of interest (in a multiple scenario reftable), as you previously mentioned.

**Comment by aestoup** [comment]
Comment for Global / local error-accuracy:

You are 75% correct.

Global and local error for model choice = nothing new.

Nothing really new too for global error/accuracy for parameter estimation (note that we focus on MSE and NMAE).

...New stat tricks = actually (only/mainly) for local posterior error/accuracy for parameter estimation. This is where Louis had to implement some new code in abcrf.

**Comment by fradav** [comment]
Ok from what I understand reading the paper there is two "local" posterior errors : 

![gif 1](https://user-images.githubusercontent.com/701587/60727017-34d97a00-9f3d-11e9-87c1-5ef0ce633b64.gif)

This is actually the variance, nothing new then.

![gif 2](https://user-images.githubusercontent.com/701587/60727107-681c0900-9f3d-11e9-8528-5dd9d6a99f22.gif)

Ok this is new, I should call this "local mae" ? I'll add it to the estimparam output.

Anything else ?
